### ISSUE TITLE
How to Git?

### ISSUE THREAD

**Issue by veqryn** [triplea-game/triplea#26]
Ok, so I do not really know how to use Git all that well.

Questions so far:

1. If I have write permission to the master branch, but I want someone to review my code before I commit it, how should I do this?

2. Branches vs Forking?  What is the difference?  Which is preferable and when to use each? And how do I keep my local branch or local fork up to date with the master?

3. When using branches, after your pull request gets accepted and merged, you want to code a new feature or something else.  What should you do?  Is it preferable to throw away the branch and start a new branch based off master?  Or should you pull master into your branch and merge it, then code from there?

4. How do I commit things under a different username/email than what my global git settings say.  No, I don't want to change my git config, because I really do have multiple users (one for work, one for personal).

Any other tips are appreciated.

**Comment by DanVanAtta** [comment]
Very briefly, to give you some quick answers (i'm going to start on a git work flows wiki)

1. pull request, the only other person that can merge it is likely redrum, but others can view and comment on it

2. forking creates a new repository in github. Think of it as a branch that exists in github. Generally the workflow is fork, clone locally, create branch locally. push local branch to fork. Do pull request of branch in your fork to the master repository

Branches in the master repository will likely be stabilization branches for releases. Test master very thoroughly, cut branch, build artifacts using that branch, publish the artifacts and do release upgrade stuff necessary. Note here, continuous build automates those steps, means we can focus on test. For example, we could pick a version and use it for thorough testing it, and then label it a stable version while we continue to build "unstable" versions continuously (but only building if our tests pass, which hopefully will mean a lot one day).

3. What you want to avoid are many branches, or branches that are "long lived". IE: branches ideally would get merged back in on a pretty frequent basis. With that said, before merging to master ideally as well the branch being merged in will be current with master. (ie: git fetch <remote>; git merge <remote>/master). Once the branch to merge is current with master, plus the additional changes, it is then merged to master. At that point you can continue working on it if you choose. For some branches that people would want to integrate with often without would get re-used. For example if a team of 3 were working on a bigger feature and needed to integrate.

But, if you have any outstanding pull requests, one thing to know is that pull requests bring in full branches. If you update the branch, the pull request is also updated.

To mention it here, branch merges work like this:
For example, branch A is merged into Branch B
Git will pick a common ancestor
Then from there gits on branch A and branch B are interleaved (using the commit timestamp).
Then a merge commit is put on top of that. The merge commit can change anything and everything, but a standard merge commit roughly generally acts as placeholder for the branch being merged in.

4. Stack overflow better resource than I (anyone else know)

**Comment by DanVanAtta** [comment]
Quick and somewhat dirty write up of the main git work flows:
https://github.com/triplea-game/triplea/wiki/Git-tips-and--Workflows

**Comment by veqryn** [comment]
1. Do we prefer forks or branches for changes?  Which should I be using?

3. You said that changing a branch changes a pull request.  What if the pull request is already merged, then I change my branch after that?  Does that make a new pull request, or wait til I make a new pull request?  Also I didn't get a clear answer on whether I should delete my branch after a pull request is merged, or just keep editing the same branch.

5. When to use rebase vs merge?

**Comment by gaborbernat** [comment]
1. Core developers (which at the moment are two) could probably be using branches. Others could fork and pull request. 

2. If we use feature branches, you can delete once merged. A [good read around the subject](http://nvie.com/posts/a-successful-git-branching-model/).
3. I prefer to always use rebase. Merges can really easily lead to a merge madness, with countless parallel codes, breaking a linear history format. I should emphasize here that we should use feature branches, if a feature is done condense it to just 2-3 commits. Merge via rebase just these 2-3 commits that are easy to read inside the history. (this way you can have 100 commits during the development, and at  the end  just squash these into 3).

**Comment by DanVanAtta** [comment]
@gaborbernat It seems the preference is for multiple small commits to keep diff's small.  If we squash, is there a way we can still have code reviews look at the tiny commits before the squash?

**Comment by gaborbernat** [comment]
I think you misunderstood. Ron said that we should keep pull requests small as possible for ease of reviewing. 

In this it does help to have feature branches and have pull request for each feature. In this case you can continue working on something else while the feature branch does get accepted (merged in, at which point you can delete the feature branch).  If a feature has modifications in 3 classes, then yes it does make sense to have it in a single or at most 2-3 commits. Because we review features, which is a functionality. And it does make sense to have it all in single place (hence the github, reviewable.io shows all modifications for a pull request, not per commit), for ease of overview.

For refactoring it can help if you have each type of refactor in a single pull request. If there are 10 refactoring types in a single pull request it's harder to review.

**Comment by DanVanAtta** [comment]
Sorry to make you spell it out, @gaborbernat , though I think that was helpful and my understanding has now changed and is more in line. Notice for example the splicing of my a recent pull request into many small ones.

I've a lot of comments about the two branch model (successful branching model link above: http://nvie.com/posts/a-successful-git-branching-model/). I don't think I like how this model played out over a year with a few dozen developers in my past experience. The feature branch concept is useful, as are other concepts discussed in that article, but keeping around two branches and release branches tends to convoluted. I can see it working very well for 1000+ engineers, but not until you hit maybe 250 or 500 would i really consider that (but it is an interesting read for certain)

I think these strategies will be more appropriate for us:
https://secure.phabricator.com/book/phabflavor/article/recommendations_on_revision_control/
https://secure.phabricator.com/book/phabflavor/article/recommendations_on_branching/


As far as how we decide on a branching model. I find that having work flows spelled out command by command are useful. There is no confusion, there is no huge learning curve for new people, there is pretty strong consistency. So I think a lot of how-to-git efforts should be focused on building a good git work-flow wiki (which has been started I'll note!)

**Comment by ron-murhammer** [comment]
+1 for the more concise pull requests. Now only if I had time to actually review stuff :/

**Comment by DanVanAtta** [comment]
For question 4 (multiple names/emails) @veqryn , looks like this may have an answer for you: https://orrsella.com/2013/08/10/git-using-different-user-emails-for-different-repositories/

**Comment by DanVanAtta** [comment]
Good to close?

**Comment by veqryn** [comment]
When I look back through commit history, I will see the same change getting done multiple times.
Why?  And how to stop this from happening?
It makes it very hard to see what actually changed in each commit, if they are not atomic and separate.

**Comment by DanVanAtta** [comment]
Can you be more specific in terms of how you are looking through the commit history?

**Comment by DanVanAtta** [comment]
If you're talking about the PR's on github, I need to do some work to rebase them. Just finished with #66 for example.

**Comment by DanVanAtta** [comment]
To be specific about what that work was, I needed to take my branches and merge them with master. At that point git saw the commits as the same and moved the common ancestor up, so then you only see the commit that adds stuff, and the merge commit that updated the branch.

Should we close this issue out? Is the wiki page on the git work flows sufficient?

**Comment by DanVanAtta** [comment]
Learned something recently myself. To update between two local branches pull can be used instead of merge. instead of: `git merge master`, can do: `git pull --rebase . master`

I finished a few moments ago going through my open pull requests and removing the unnecessary merge commits by rebasing instead of merge. 

I also found some interesting reading about merge vs rebase, general git stuff, and a useful tool for working with git called "hub":
http://blog.differential.com/best-way-to-merge-a-github-pull-request/
http://mislav.uniqpath.com/2013/02/merge-vs-rebase/
http://mislav.uniqpath.com/2010/07/git-tips/
https://hub.github.com/

**Comment by veqryn** [comment]
So, right now in the branches tab, GitHub says that release_1_8_0_7 is 2 commits behind master.
Previously, it said it was 1 commit behind master.
I decided I would merge changes in master, into the release_1_8_0_7 branch, that way they would be equal. I followed instructions online that recommended rebase, so I did this, while on a local checkout of the release_1_8_0_7 branch:
git fetch origin
git pull
git rebase master
git pull
git push

The result of this, is that there are 2 new commits in our history that appear to do nothing, and the release_1_8_0_7 branch is now 2 commits behind master (https://github.com/triplea-game/triplea/commit/c2a681ac3ba62e5dc85d1e1fa172ec71bd104831 and https://github.com/triplea-game/triplea/commit/e48b2e01d69ecfae3212c1b68012fa179f84b1b3).
How is this even possible?

Also, why does git show changes to files in pull requests, when the files have already changed to become that?  For example, you created two pull requests to revert the mac line endings change, one for master and one for the release branch.  I merged the release branch, then created a pull request for the release branch to master and merged that.  So now master has the correct line endings.  But when I open up your second pull request, it shows the change still there, instead of showing something like "no changes", or "invalid pull request because this stuff has already happened".

Git is really frustrating me here!

**Comment by DanVanAtta** [comment]
The release branch is meant to get behind from master. As I think you may know, the commits currently on master but not now on release are going into the next release.

So, about the two useless commits:
- `git pull` by default adds a commit. You want instead `git pull --rebase`. There is a way to set this a global property. An alias of "gp" or "git_pull" could be warranted. Or perhaps create a ~/bin/git_pull, that will also do the git fetch for you.

- To move the commits from the release branch into master, we should not be touching the release branch. So you should have master checked out. Then:

```
git fetch origin
git checkout master
git pull --rebase 
git pull --rebase origin <release_branch_name>
git push origin master
```

A second time with explanatory comments and some verification you can do along the way:
```
git fetch origin
git checkout master
git pull --rebase  ## make sure our local master is  up to date
git log  ## verify log has latest master and not the release commits
git pull --rebase origin <release_branch_name>
git log   # last commits should be from the release branch, added on top of what was previously there
git push origin master  # update remote
git remote -v  ## to see what we just updated (whether fork or the main repo)
```

**Comment by veqryn** [comment]
I get that the release branch is supposed to fall behind.  But as of right now, they should be equal because the only changes in either are all just fixes for the release_1_8_0_7 branch.  This is why I wanted them equal, because I didn't understand why they weren't (and I have no way to compare and see what the differences are).

Secondly, I wanted to merge the changes in master, into the release branch, not the other way around.  Is this what your code block does, or is it backwards?

**Comment by DanVanAtta** [comment]
Backwards. I was hoping with the dual PR to both branches we would avoid this. Regardless, your original code block then was pretty close:
```
git checkout <release_branch>
git fetch origin
git pull --rebase origin master
git push origin <release_branch>
```

If you want to "undo" the extra commits, and then redo the above:
```
git log
git reset --soft HEAD^  # repeat this until git log shows the last commit is the last good one
git status  ## see what changed, you'll now work to clear the status

git reset <file> && git checkout <file>   # where file is anything showing as modified or moved in the git status, this will undo changes brought in, keep doing this until the git status is clean. Don't worry about untracked files.

<now redo the pull request stuff>

git push origin -f <release_branch>  # note, the -f is to "force". Basically we rewrote the history when we did the reset, this caused a merge conflict, so we are overriding that. The '-f' should be used with care.
```

**Comment by DanVanAtta** [comment]
Note in the above, if you want to "abort" the redo after you've done a reset, a "git checkout <branch_name>" should be sufficient to move your commit pointer back to the tip of the branch.

**Comment by DanVanAtta** [comment]
More to discuss? Can we close? If someone references this issue it will be painfully long to read. I recommend we close and open new tickets on focused topics. LIke, how to git for branching, or how to git for release management.

**Comment by DanVanAtta** [comment]
@veqryn Found an answer to this:
"How do I commit things under a different username/email than what my global git settings say. No, I don't want to change my git config, because I really do have multiple users (one for work, one for personal)."

You can set per repository settings with git for windows. From right click menu >  Git GUI > edit menu bar, "options":
![per_repo_setting](https://cloud.githubusercontent.com/assets/12397753/10231791/8b79537e-6839-11e5-802e-945d1f1c89c8.png)


With CLI, you can do it per repository by simply dropping the global flag, ie:
`git config user.name "John Doe"`

The per repo value is then stored in `.git/config`:
```
$ tail -2 .git/config 
[user]
	name = DanVanAtta
```
